scheduler/test/cook/test/simulator.clj

```diff
@@ -0,0 +1,571 @@
+                                                 :match-trigger-chan matcher-trigger-chan
+                                                 :rebalancer-trigger-chan rebalancer-trigger-chan
+                                                 ;; Don't care about these yet
+                                                 :straggler-trigger-chan (async/chan)
+                                                 :lingering-task-trigger-chan (async/chan)
+                                                 :cancelled-task-trigger-chan (async/chan)}})]
+    ;; We are setting time to enable us to have deterministic runs
+    ;; of the simulator while hooking into the scheduler as non-invasively
+    ;; as possible. A longer explanation can be found in the simulator dev docs.
+    (org.joda.time.DateTimeUtils/setCurrentMillisFixed initial-time)
+    (log/info "Starting simulation.")
+    (with-cook-scheduler mesos-datomic-conn make-mesos-driver-fn
+      scheduler-config
+      (try
+        (doseq [{:keys [user mem cpus gpus]} (:shares config)]
+          (share/set-share! mesos-datomic-conn user "simulation" :mem mem :cpus cpus :gpus gpus))
+        (loop [trace trace
+               simulation-time simulation-time
+               fake-real-time initial-time]
+
+          (org.joda.time.DateTimeUtils/setCurrentMillisFixed fake-real-time)
+
+          (let [start-ms (System/currentTimeMillis)
+                submission-batch (take-while #(<= (:submit-time-ms %) simulation-time) trace)
+                send-offers-complete-chan (async/chan)
+                flush-complete-chan (async/chan)
+                match-complete-chan (async/chan)
+                rank-complete-chan (async/chan)
+                rebalancer-complete-chan (async/chan)]
+            (log/info "Simulation time: " simulation-time)
+            (log/info "submission-batch size: " (count submission-batch))
+
+            (log/info "Submitting batch")
+            ;; Submit new jobs
+            (doseq [job submission-batch]
+              (org.joda.time.DateTimeUtils/setCurrentMillisFixed (inc (.getTime (tc/to-date (t/now)))))
+              (submit-job mesos-datomic-conn job))
+            ;; Ensure peer has acknowledged the new jobs
+            (when (seq submission-batch)
+              (poll-until #(d/q '[:find ?j .
+                                      :in $ ?t
+                                      :where
+                                      [?j :job/submit-time ?t]]
+                                    (d/db mesos-datomic-conn)
+                                    ;; This relies on
+                                    ;; 1. Time is controlled
+                                    ;; 2. We increment time for each job submitted
+                                    (tc/to-date (t/now)))
+                        50
+                        60000))
+            (log/info "Batch submission complete")
+
+            ;; Rebalance
+            (log/info "Starting rebalance")
```

Comment:
Yea, that would be better but would mean more things to configure for the sim (cycle step time, rebalancer interval, rank interval, match interval...). Essentially as written, all the process intervals equal the cycle step time.